This file is based on the ETL and Data Cleaning done by Sean

In [25]:
import warnings
warnings.filterwarnings('ignore')

In [26]:
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter

In [27]:
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

ModuleNotFoundError: No module named 'imblearn'

# Read the CSV and Perform Basic Data Cleaning

In [28]:
# Read csv into pandas dataframe
main_df = pd.read_csv('../../Data/Main_Sean_File_Numeric_2019-2020 NSCH_Topical_CAHMI_DRC_1920.csv')
main_df.head()

age3_1920  age5_1920  sex_1920  hispanic_1920  race4_1920  raceASIA_1920  \
0          3          5         1              2           2              2   
1          1          2         2              2           2              2   
2          3          4         1              2           2              2   
3          3          5         2              2           2              2   
4          1          2         2              2           3              3   

   race7_1920  PrntNativity_1920  HHLanguage_1920  hisplang_1920  ...  \
0           2                  1                1              3  ...   
1           2                  1                1              3  ...   
2           2                  1                1              3  ...   
3           2                 99                1              3  ...   
4           3                  1                1              3  ...   

   npm9bullied_1920  npm9bully_1920  npm10PrVi_1920  npm11MHCSHCN_1920  \
0                 1               1               2                 95   
1                90              90              95                 95   
2                 2               1               1                 95   
3                 2               1               1                 95   
4                90              90              95                  2   

   npm11MHnonCSHCN_1920  npm12TransCSHCN_1920  npm12TransNonCSHCN_1920  \
0                     1                    95                        2   
1                     1                    90                       90   
2                     1                    95                        2   
3                     1                    95                        1   
4                    95                    90                       90   

   npm13_2DePrev_1920  npm14_2Smoke_1920  npm15AdIns_1920  
0                   1                  2                1  
1                   1                  2                1  
2                   1                  2                2  
3                   1                  2                2  
4                   1                  2                1  

[5 rows x 387 columns]

In [35]:
# Read csv into pandas dataframe
original_raw_df = pd.read_csv('../../Data/2019-2020 NSCH_Topical_CAHMI_DRC_Original.csv')
original_df.head()

NameError: name 'original_df' is not defined

In [38]:
# Read csv into pandas dataframe
refference_df = pd.read_csv('../../Data/RAW_2019-2020 NSCH_Topical_CAHMI_DRC.csv, encoding="latin1"'')
refference_df.head()

SyntaxError: EOL while scanning string literal (Temp/ipykernel_12676/3423488910.py, line 2)

In [ ]:
# Combine the data into a single dataset
merge_df = pd.merge(main_df, school_data_df, how="left", on=["school_name", "school_name"])
school_data_complete_df.head()

Are K2Q35A & K2Q35B mutulally exclusive?  What are the total counts of each?
K2Q35A Question: Has a doctor or other health care provider EVER told you that this child has… Autism or Autism Spectrum Disorder (ASD)? Include diagnoses of Asperger’’s Disorder or Pervasive Developmental Disorder (PDD).

K2Q35B Question: If yes, does this child CURRENTLY have the condition? (variable name differs based on condition)

In [15]:
Autism_types_df = main_df.values['K2Q35A', 'K2Q35B']
Autism_types_df.head()

#total_budget = school_data_df["budget"].sum()

IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices

In [ ]:
# pandas.DataFrame.transpose(NSCH_2019_2020-df)

In [ ]:
This particular analysis will look at one subsection of the survey - 

In [16]:
# https://www.childhealthdata.org/browse/survey
# C:/Users/tligh/Documents/MSU/Challenges/Sensory_Needs_Occupational_Therapy/data/NSCH_Data_Set_Multiple_Years_Tara.csv

# K2Q35A = Has a doctor or other health care provider EVER told you that this child has… Autism or Autism Spectrum Disorder (ASD)? 
# Include diagnoses of Asperger’’s Disorder or Pervasive Developmental Disorder (PDD).
columns = [

]

target = ["K2Q35A"]

In [5]:
# Load the data
file_path = Path('C:/Users/tligh/Documents/MSU/Challenges/Sensory_Needs_Occupational_Therapy/data/NSCH_Data_Set_Multiple_Years_Tara.csv')
df = pd.read_csv(file_path, skiprows=1)[:-2]
df = df.loc[:, columns].copy()

# Drop the null columns where all values are null
df = df.dropna(axis='columns', how='all')

# Drop the null rows
df = df.dropna()

# Remove the `Issued` loan status
issued_mask = df['loan_status'] != 'Issued'
df = df.loc[issued_mask]

# convert interest rate to numerical
df['int_rate'] = df['int_rate'].str.replace('%', '')
df['int_rate'] = df['int_rate'].astype('float') / 100


# Convert the target column values to low_risk and high_risk based on their values
x = {'Current': 'low_risk'}   
df = df.replace(x)

x = dict.fromkeys(['Late (31-120 days)', 'Late (16-30 days)', 'Default', 'In Grace Period'], 'high_risk')    
df = df.replace(x)

df.reset_index(inplace=True, drop=True)

df.head()

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0    10500.0    0.1719       375.35           RENT     66000.0   
1    25000.0    0.2000       929.09       MORTGAGE    105000.0   
2    20000.0    0.2000       529.88       MORTGAGE     56000.0   
3    10000.0    0.1640       353.55           RENT     92000.0   
4    22000.0    0.1474       520.39       MORTGAGE     52000.0   

  verification_status   issue_d loan_status pymnt_plan    dti  ...  \
0     Source Verified  Mar-2019    low_risk          n  27.24  ...   
1            Verified  Mar-2019    low_risk          n  20.23  ...   
2            Verified  Mar-2019    low_risk          n  24.26  ...   
3            Verified  Mar-2019    low_risk          n  31.44  ...   
4        Not Verified  Mar-2019    low_risk          n  18.76  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            85.7             100.0                   0.0        0.0   
1            91.2              50.0                   1.0        0.0   
2            66.7              50.0                   0.0        0.0   
3           100.0              50.0                   1.0        0.0   
4           100.0               0.0                   0.0        0.0   

   tot_hi_cred_lim  total_bal_ex_mort total_bc_limit  \
0          65687.0            38199.0         2000.0   
1         271427.0            60641.0        41200.0   
2          60644.0            45684.0         7500.0   
3          99506.0            68784.0        19700.0   
4         219750.0            25919.0        27600.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     61987.0              N                     N  
1                     49197.0              N                     N  
2                     43144.0              N                     N  
3                     76506.0              N                     N  
4                     20000.0              N                     N  

[5 rows x 86 columns]

# Split the Data into Training and Testing

In [6]:
# Create our features
X = df.drop(columns='loan_status')
X = pd.get_dummies(X)

# Create our target
y = df.loc[:, target].copy()

In [7]:
X.describe()

loan_amnt      int_rate   installment    annual_inc           dti  \
count  68817.000000  68817.000000  68817.000000  6.881700e+04  68817.000000   
mean   16677.594562      0.127718    480.652863  8.821371e+04     21.778153   
std    10277.348590      0.048130    288.062432  1.155800e+05     20.199244   
min     1000.000000      0.060000     30.890000  4.000000e+01      0.000000   
25%     9000.000000      0.088100    265.730000  5.000000e+04     13.890000   
50%    15000.000000      0.118000    404.560000  7.300000e+04     19.760000   
75%    24000.000000      0.155700    648.100000  1.040000e+05     26.660000   
max    40000.000000      0.308400   1676.230000  8.797500e+06    999.000000   

        delinq_2yrs  inq_last_6mths      open_acc       pub_rec  \
count  68817.000000    68817.000000  68817.000000  68817.000000   
mean       0.217766        0.497697     12.587340      0.126030   
std        0.718367        0.758122      6.022869      0.336797   
min        0.000000        0.000000      2.000000      0.000000   
25%        0.000000        0.000000      8.000000      0.000000   
50%        0.000000        0.000000     11.000000      0.000000   
75%        0.000000        1.000000     16.000000      0.000000   
max       18.000000        5.000000     72.000000      4.000000   

           revol_bal  ...  issue_d_Mar-2019  pymnt_plan_n  \
count   68817.000000  ...      68817.000000       68817.0   
mean    17604.142828  ...          0.177238           1.0   
std     21835.880400  ...          0.381873           0.0   
min         0.000000  ...          0.000000           1.0   
25%      6293.000000  ...          0.000000           1.0   
50%     12068.000000  ...          0.000000           1.0   
75%     21735.000000  ...          0.000000           1.0   
max    587191.000000  ...          1.000000           1.0   

       initial_list_status_f  initial_list_status_w  next_pymnt_d_Apr-2019  \
count           68817.000000           68817.000000           68817.000000   
mean                0.123879               0.876121               0.383161   
std                 0.329446               0.329446               0.486161   
min                 0.000000               0.000000               0.000000   
25%                 0.000000               1.000000               0.000000   
50%                 0.000000               1.000000               0.000000   
75%                 0.000000               1.000000               1.000000   
max                 1.000000               1.000000               1.000000   

       next_pymnt_d_May-2019  application_type_Individual  \
count           68817.000000                 68817.000000   
mean                0.616839                     0.860340   
std                 0.486161                     0.346637   
min                 0.000000                     0.000000   
25%                 0.000000                     1.000000   
50%                 1.000000                     1.000000   
75%                 1.000000                     1.000000   
max                 1.000000                     1.000000   

       application_type_Joint App  hardship_flag_N  debt_settlement_flag_N  
count                68817.000000          68817.0                 68817.0  
mean                     0.139660              1.0                     1.0  
std                      0.346637              0.0                     0.0  
min                      0.000000              1.0                     1.0  
25%                      0.000000              1.0                     1.0  
50%                      0.000000              1.0                     1.0  
75%                      0.000000              1.0                     1.0  
max                      1.000000              1.0                     1.0  

[8 rows x 95 columns]

In [8]:
# Check the balance of our target values
y['loan_status'].value_counts()

low_risk     68470
high_risk      347
Name: loan_status, dtype: int64

In [9]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

# Ensemble Learners

In this section, you will compare two ensemble algorithms to determine which algorithm results in the best performance. You will train a Balanced Random Forest Classifier and an Easy Ensemble AdaBoost classifier . For each algorithm, be sure to complete the folliowing steps:

1. Train the model using the training data. 
2. Calculate the balanced accuracy score from sklearn.metrics.
3. Print the confusion matrix from sklearn.metrics.
4. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.
5. For the Balanced Random Forest Classifier onely, print the feature importance sorted in descending order (most important feature to least important) along with the feature score

Note: Use a random state of 1 for each algorithm to ensure consistency between tests

### Balanced Random Forest Classifier

In [10]:
# Resample the training data with the BalancedRandomForestClassifier
from imblearn.ensemble import BalancedRandomForestClassifier
brf = BalancedRandomForestClassifier(n_estimators=100, random_state=1)
brf.fit(X_train, y_train)

BalancedRandomForestClassifier(random_state=1)

In [11]:
# Calculated the balanced accuracy score
y_pred = brf.predict(X_test)
balanced_accuracy_score(y_test, y_pred)

0.7885466545953005

In [12]:
# Display the confusion matrix
confusion_matrix(y_test, y_pred)

array([[   71,    30],
       [ 2153, 14951]])

In [13]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.03      0.70      0.87      0.06      0.78      0.60       101
   low_risk       1.00      0.87      0.70      0.93      0.78      0.62     17104

avg / total       0.99      0.87      0.70      0.93      0.78      0.62     17205



In [14]:
# Get importances and features
importances = brf.feature_importances_
cols = X.columns

# Store in a DataFrame
feature_importances_df = pd.DataFrame({'feature':cols, 'importance': importances})
feature_importances_df.head()

feature  importance
0    loan_amnt    0.010580
1     int_rate    0.029670
2  installment    0.019802
3   annual_inc    0.014833
4          dti    0.017471

In [15]:
# Sort the DataFrame
feature_importances_df.sort_values('importance', ascending=False)

feature  importance
15          total_rec_prncp    0.078768
13              total_pymnt    0.058838
14          total_pymnt_inv    0.056256
16            total_rec_int    0.053555
20          last_pymnt_amnt    0.050033
..                      ...         ...
19  collection_recovery_fee    0.000000
22              policy_code    0.000000
23           acc_now_delinq    0.000000
46              delinq_amnt    0.000000
94   debt_settlement_flag_N    0.000000

[95 rows x 2 columns]

### Easy Ensemble AdaBoost Classifier

In [16]:
# Train the EasyEnsembleClassifier
from imblearn.ensemble import EasyEnsembleClassifier
eec = EasyEnsembleClassifier(n_estimators=100, random_state=1)
eec.fit(X_train, y_train)

EasyEnsembleClassifier(n_estimators=100, random_state=1)

In [17]:
# Calculated the balanced accuracy score
y_pred = eec.predict(X_test)
balanced_accuracy_score(y_test, y_pred)

0.9316600714093861

In [18]:
# Display the confusion matrix
confusion_matrix(y_test, y_pred)

array([[   93,     8],
       [  983, 16121]])

In [19]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.09      0.92      0.94      0.16      0.93      0.87       101
   low_risk       1.00      0.94      0.92      0.97      0.93      0.87     17104

avg / total       0.99      0.94      0.92      0.97      0.93      0.87     17205



### Summary and Analysis of Results

The random forest classifier, with and without AdaBoost, failed to achieve useable performance. The balanced random forest classifier's precision is 0.04, meaning that in 100 loan applications that were flagged to be bad, only 4 were actually bad loan applications. The model's recall/sensitivity is 0.67, meaning that it detected 67% of bad loan applications. The F1 score is low at 0.07, since either a low precision or recall will result in a lower F1 score.

The random forest classifier with AdaBoost, while achieving better results, still suffered from inadequate predictive power. Its precision score is 0.09 and its recall 0.92. The F1 score, again, is skewed low at 0.16 by the low precision score.

The performances of both models are insufficient for commercial application.